sql: restore sanity in the handling of special column names

knz opened the pull request:
Fixes  #16813.
Fixes  #16814.

This should make it into the next 1.0.x, do you agree?

knz committed:
sql: add a missing test

The fix for a name escaping bug in
4c044767a141d97094d24ce2497b92e80f0e4a4e was missing a test. This
patch adds it and also checks that other things in a table definition
are properly escaped when printed out.

knz committed:
sql: handle special characters in view column names

Prior to this patch, view descriptors would not be properly validated
if the source table's columns had names containing special characters.

This patch alleviates the issue by enforcing the proper escaping
mechanism.

Fixes #16813.

knz committed:
sql: clean up IndexedVars, fix bug in CREATE TABLE AS

This patch changes the code that binds IndexedVars to a container to
ensure that only freshly-allocated IndexedVars are modified in-place.
The other container-less use of IndexedVars, namely ordinal references
of the form @N in the input syntax, are not any more modified
in-place; the BindIfUnbound method now returns a freshly allocated
object for those.

This in turn ensures that expressions coming from statements
containing ordinal references are not modified in-place, and become
suitable for multiple processing by analyzeExpr in different plan
contexts (previously, it was safe only when the same IndexedVar was
used in multiple expressions in the same plan context).

This in turns fixes a panic inducing bug in the handling of CREATE
TABLE AS, when the source select contained ordinal references.

Fixes #16814.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16815)
<!-- Reviewable:end -->

knz commented:
tests fixed, ptal

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I'd feel better if @raduberinde also had a look before merging.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 5 of 5 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KnuzLHmRo6SgQfQVaGL:bobh8e8)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/sql/parser/indexed_vars.go:
```diff
@@ -238,3 +254,19 @@ func (h *IndexedVarHelper) VisitPre(expr Expr) (recurse bool, newExpr Expr) {
+
+type unboundContainerType struct{}
+
+var unboundContainer = &unboundContainerType{}
```
Please add a comment explaining when this is used, and how it differs from a `nil` container. I think the gist of it is that it indicates that we were previously bound to a container but we aren't anymore, right?

RaduBerinde commented on pkg/sql/parser/indexed_vars.go:
```diff
@@ -111,35 +122,42 @@ type IndexedVarHelper struct {
+// BindIfUnbound ensures the IndexedVar is attached to this helper's container.
+// - for freshly created IndexedVars (with a nil container) this will bind in-place.
+// - for already bound IndexedVar, bound to this container, this will return the same ivar unchanged.
+// - for ordinal references (with an explicit unboundedContainer) this will return a new var.
```
unboundContainer

RaduBerinde reviewed (commented):
LGTM, some minor comments.
I do think it should make it in a 1.0.x release, but I am a bit concerned at the amount of code touched. Maybe we want it on master for a while before doing that.

bdarnell commented:
👍 for inclusion in 1.0.4. 1.0.3 is going out this week, so we'll have this on master for a couple of weeks before we cut the next release.

tamird commented:
> This it turns fixes a panic inducing bug in the handling of CREATE
> TABLE AS, when the source select contained ordinal references.

s/it turns/in turn/

Also, can you put each "Fixes #xxx" in the commit that fixes it? Looks like each issue is fixed by one commit, but please correct me if I'm wrong.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-Ko9u-ap8I4X79ul3V5T:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we want to check this assumption? (panic if we return anything other than `ivar`)
</blockquote></details>

+1

---

*[pkg/sql/logictest/testdata/logic_test/name_escapes, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9sPfa6lz75ss6z7ok:-Ko9sPfa6lz75ss6z7ol:b-6kfxfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b25008a2c76fb7bb42a54f6c2e5b1b8b6c1c960/pkg/sql/logictest/testdata/logic_test/name_escapes#L41)):*
> ```
> CREATE TABLE unused (x ")
> ) INTERLEAVE IN PARENT "woo; DROP USER humpty;" (x)
> ```

errant empty newline

looks like r2 "fixed" it

---

*[pkg/sql/parser/indexed_vars.go, line 107 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9uf6VLiKi-zC-bB7t:-Ko9uf6VLiKi-zC-bB7u:bm0wn7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L107)):*
> ```Go
> // NewIndexedVar is a helper routine to create a standalone Indexedvar
> // with the given index value. This needs to undergo BindIfUnbound()
> // below before it can be fully used.  The difference with ordinal
> ```

double space

---

*[pkg/sql/parser/indexed_vars.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9v-bO_7jeBbrqVxaf:-Ko9v-bO_7jeBbrqVxag:b-1ldjzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L109)):*
> Do not use NewIndexedVar for AST nodes

Consider giving this its own paragraph.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-Ko9tbexbQnmJPkbwoqd:b8yhgdj)*
<!-- Sent from Reviewable.io -->

knz commented:
I made the proposed changes.

I don't usually indicate which commit fixes which issue. Instead I explain (and repeat) what the issue is in the commit message that fixes each issue. This ensures that the git history can be read and understood without the github context. The relationship between PRs and issues is a github thing. 

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoDTuL3dfrB2A0CLjVu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/expr.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG0-r3-44:-KoDTwDsT-Nq2M4R-Z6S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/distsqlrun/expr.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think we should return `false, expr` in the error case, or we may screw up the expression.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/name_escapes, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9sPfa6lz75ss6z7ok:-KoDUWxdH6mtihpFPkg3:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b25008a2c76fb7bb42a54f6c2e5b1b8b6c1c960/pkg/sql/logictest/testdata/logic_test/name_escapes#L41)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errant empty newline

looks like r2 "fixed" it
</blockquote></details>

Fixed.

---

*[pkg/sql/parser/indexed_vars.go, line 107 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9uf6VLiKi-zC-bB7t:-KoDUgEhR8K-ogdUEFvG:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L107)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

double space
</blockquote></details>

fixed

---

*[pkg/sql/parser/indexed_vars.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9v-bO_7jeBbrqVxaf:-KoDUhejdFhN9Kb1iSK-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> Do not use NewIndexedVar for AST nodes

Consider giving this its own paragraph.
</blockquote></details>

Done.

---

*[pkg/sql/parser/indexed_vars.go, line 128 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG1-r3-128:-KoDUkK6vxuaTiwURqPh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L128)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

unboundContainer
</blockquote></details>

Done.

---

*[pkg/sql/parser/indexed_vars.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG1-r3-260:-KoDVAmysCjzbUrFdX24:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L260)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Please add a comment explaining when this is used, and how it differs from a `nil` container. I think the gist of it is that it indicates that we were previously bound to a container but we aren't anymore, right?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoDVMzAvOJVWd9GNf3k:b-mkcdoc)*
<!-- Sent from Reviewable.io -->

tamird commented:
> I don't usually indicate which commit fixes which issue. Instead I explain (and repeat) what the issue is in the commit message that fixes each issue. This ensures that the git history can be read and understood without the github context. The relationship between PRs and issues is a github thing.

I agree that explaining the issue in the commit message is helpful, but what's the downside of _also_ including the issue reference? Yes, it's a github thing, but since the atomic unit of change that fixes the issue is the commit, it seems strictly beneficial to include.

---

Reviewed 12 of 12 files at r4, 8 of 8 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoIM4VH418p4dPye-YB:b-ynzdvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The comment above should be updated now; also, if you don't mind splitting the two cases:

```
			if newIvar, err := aggHelper.BindIfUnbound(ivar); err != nil {
				panic(err)
			} else if newIvar != ivar {
				panic(fmt.Sprintf("unexpected binding: %v, expected: %v", newIvar, ivar"))
			}
```

---

*[pkg/sql/parser/indexed_vars.go, line 111 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KoIMZ_JEGrpSCJB9ObH:-KoIMZ_JEGrpSCJB9ObI:b-xzd9xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1752e18a929aff0e2786436c28d9db788b9b2e90/pkg/sql/parser/indexed_vars.go#L111)):*
> ```Go
> // in-place by BindIfUnbound.
> //
> // Do not use NewIndexedVar for AST nodes
> ```

re-wrap this plz

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoILeBqZZVTNVycnE15:b3ces3u)*
<!-- Sent from Reviewable.io -->

knz commented:
okay, that works for me; I changed the commit messages.

---

Review status: 0 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoJPq2mSjY0Jy32IZOL:bt4ox15) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The comment above should be updated now; also, if you don't mind splitting the two cases:

```
			if newIvar, err := aggHelper.BindIfUnbound(ivar); err != nil {
				panic(err)
			} else if newIvar != ivar {
				panic(fmt.Sprintf("unexpected binding: %v, expected: %v", newIvar, ivar"))
			}
```
</blockquote></details>

Done. Thanks for the suggestion!

---

*[pkg/sql/parser/indexed_vars.go, line 111 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KoIMZ_JEGrpSCJB9ObH:-KoJPerPLQ_OAuYMP1cA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1752e18a929aff0e2786436c28d9db788b9b2e90/pkg/sql/parser/indexed_vars.go#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

re-wrap this plz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoJP5DmgSZ0PcAeGStX:bog1gp5)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

---

Reviewed 7 of 8 files at r8, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoJQiJ9IxyxwzZumzFt:brfqlux)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!
